### PR TITLE
Use XDG_CACHE_HOME for locating the build cache directory

### PR DESCRIPTION
**Pull request by bombela**
Follows: https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html

In short, use $XDG_CACHE_HOME if defined, else $HOME/.cache/

### CODE REVIEW

File: build/base.mk [hunk 1]
```diff
@@ -1,7 +1,9 @@
 UNAME_OS := $(shell uname -s)
 UNAME_ARCH := $(shell uname -m)
 
-CACHE_BASE := $(HOME)/.yarpc-go
+XDG_CACHE_HOME ?= $(HOME)/.cache
+
+CACHE_BASE := $(XDG_CACHE_HOME)/yarpc-go
```
Comment (peter-edge) [inline build/base.mk]: Make this `?=` so we can set it on Travis. I also don't like making `CACHE_BASE` depend on `XDG_CACHE_HOME` all the time, as this is linux specific. I know you're setting it if it's not set, but the semantics just don't make me feel great - yarpc building isn't linux specific, so if I want to override `CACHE_BASE`, I don't want to have to do it by setting `XDG_CACHE_HOME`.

File: .travis.yml [hunk 2]
```diff
@@ -3,7 +3,7 @@ language: go
 go_import_path: go.uber.org/yarpc
 cache:
   directories:
-    - $HOME/.yarpc-go
+    - $HOME/.cache/yarpc-go
```
Comment (peter-edge) [inline .travis.yml]: How do you know that `XDG_CACHE_HOME` is either not set, or equals `${HOME}/.cache`? If it is not on travis, then this will fail. I'm betting this is just set on all linux distros, but I don't like assuming this. I'd prefer we make `CACHE_BASE` optionally set in `base.mk`, and explicitly set it as a global variable https://docs.travis-ci.com/user/environment-variables/#Global-Variables

```
env:
  global:
    - CACHE_BASE=$HOME/.cache/yarpc-go
```

File: build/base.mk (hunk 1 above)
Comment (bombela) [inline build/base.mk]: XDG_CACHE_HOME is not linux specific, but freedesktop. Which includes most of not all linux distribution.

On MacOSX, from Apple documentation. https://developer.apple.com/library/content/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/FileSystemOverview/FileSystemOverview.html:

> Put data cache files in the `Library/Caches/` directory. Cache data can be used for any data that needs to persist longer than temporary data, but not as long as a support file. Generally speaking, the application does not require cache data to operate properly, but it can use cache data to improve performance. Examples of cache data include (but are not limited to) database cache files and transient, downloadable content. Note that the system may delete the Caches/ directory to free up disk space, so your app must be able to re-create or download these files as needed.

We should determine the OS, if Mac, use `Library/Caches/`. Else, use `XDG_CACHE_HOME`, else use a generic `~/.cache`.

I couldn't find the equivalent document for MS Windows. But our makefiles are GNU make anyway.

Let me know what you think.

File: .travis.yml (hunk 2 above)
Comment (bombela) [inline .travis.yml]: I effectively assumed that travis doesn't set `XDG_CACHE_HOME`. In fact XDG_CACHE_HOME is very rarely set on most distros. This is because the default "~/.cache" is perfectly reasonable. On my linux system, only non-standards settings are set via `XDG_*` env vars for example (mostly i3 config and data dirs).

But you raise a very valid point, I will use some env var.

**Comment by bombela** [comment]
I decided to use `XDG_CACHE_HOME` as the global configurable instead of `CACHE_BASE`. This is because there is way less chances of collision with `XDG_CACHE_HOME` than with `CACHE_BASE`. Also, if `XDG_CACHE_HOME` is set, it is likely to be set to some sensible value.

**Comment by peter-edge** [comment]
I don't 100% get the desire to not have collisions with `CACHE_BASE` - I see `CACHE_BASE` as something I might want to set in my `.bash_aliases`, i.e. I "want" a collision. `XDG_CACHE_HOME` feels weird to set in a `.bash_aliases` for mac, I think it makes more sense as `CACHE_BASE`.

tl;dr: I think the whole point is to have a variable that makes sense to put in your `.bash_aliases`, and `XDG_CACHE_HOME` is linux/freedesktop-specific. I rather this stay as `CACHE_BASE`.

File: .travis.yml (hunk 2 above)
Comment (peter-edge) [inline .travis.yml]: See comment about `CACHE_BASE`, but are you sure added environment variables are available here? Ie verify this on Travis CI.

File: build/base.mk (hunk 1 above)
Comment (peter-edge) [inline build/base.mk]: Note I rather this be `.cache` always unless `XDG_CACHE_HOME` is set, ie I want this to be in my homedir for mac, and not be in `Library/Caches`. This is so I have easy access to it (other things use `.cache`, I think Docker does for example).

**Comment by peter-edge** [comment]
The real tl;dr here is I'm ok if you want to have some logic saying "if `XDG_CACHE_HOME is set, use it, otherwise use `CACHE_BASE`", if you want to rely on `XDG_CACHE_HOME` for yourself, but the general case for the rest of us should be `.cache` and `CACHE_BASE` as the settable variable. Before this PR I hadn't heard of `XDG_CACHE_HOME`, and most of us are using macs.

**Comment by bombela** [comment]
But the goal of `XDG_CACHE_HOME` is to be part of your env if you ever need to. Hence why I thought it was reasonable to use it. Additionally, I would argue that we build on linux, and our production is on linux. In short, everything is designed to work on linux, not the other way around. Building on mac is only for convenience during development, so as to avoid using a virtual machine/remote dev server all the time.

 `CACHE_BASE` is also not general, as it is the base cache directory for the yarpc-go build. I can rename `CACHE_BASE` into `YARPCGO_BUILD_CACHE_BASE`. And if `YARPCGO_BUILD_CACHE_BASE` is not defined, I can fallback to `XDG_CACHE_HOME/yarpc` on linux and `~/Library/Cache` on Mac. What do you think?

**Comment by peter-edge** [comment]
These variables are only for development, and most of us use macs, and I feel like `CACHE_BASE` being non-yarpc-specific is actually good. If we were to have another repository adopt this build system, it would apply to that too, you wouldn't have to set a different variable for every repo.

If you want to do the longer variable that's ok, I just don't think `XDG_CACHE_HOME` should be what is set, I don't want that in my .bashrc.

**Comment by blampe** [comment]
`XDG_CACHE_HOME` is consistent with what we use in [yab](https://github.com/yarpc/yab/blob/dev/main_test.go#L47) and that's been working well enough for OS X.

**Comment by peter-edge** [comment]
Please just make sure the Travis cache is working properly

**Comment by peter-edge** [comment]
@blampe note that's `XDG_CONFIG_HOME` in yab, but regardless - just because yab is doing it doesn't mean it's the right thing to do - `XDG_*` is clearly a freedesktop-specific thing, and one I have never heard of before. It's really confusing to see that in my mac env. If anything, yab should change what it is doing, our development environment should not be freedesktop-specific.

I guess this isn't a big deal, but I'm surprised there's support for this change, I can unblock it though. What I do need before merging is mac does not default to `Library/Caches`, instead defaulting to `~/.cache` - this is what I would expect as a developer, I don't do any development caching in `Library`.

**Comment by peter-edge** [comment]
https://github.com/yarpc/yarpc-go/pull/1013 - if you merge this, then we're good to go.

**Comment by blampe** [comment]
We should respect the spec not overrule it http://stackoverflow.com/questions/3373948/equivalents-of-xdg-config-home-and-xdg-data-home-on-mac-os-x/5084892
